# Coursera Applied Data Science: Course 4 Week3 Part1

NOTE TO MARKER

I have added the code below I would have used to collect from Web through BeautifulSoup - commented
Due to my company firewall, I am not able to do this, so I used the alternative of read_csv


In [1]:
import pandas as pd
import requests
from bs4 import BeautifulSoup

#url_to_scrape = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
#r = requests.get(url_to_scrape)
#soup = BeautifulSoup(r.text)

In [2]:
#Create Dataframe

In [2]:
import pandas as pd
df = pd.read_csv('Wikipage.csv')
df.head(12)

Postcode           Borough     Neighbourhood
0       M1A      Not assigned      Not assigned
1       M2A      Not assigned      Not assigned
2       M3A        North York         Parkwoods
3       M4A        North York  Victoria Village
4       M5A  Downtown Toronto      Harbourfront
5       M5A  Downtown Toronto       Regent Park
6       M6A        North York  Lawrence Heights
7       M6A        North York    Lawrence Manor
8       M7A      Queen's Park      Not assigned
9       M8A      Not assigned      Not assigned
10      M9A         Etobicoke  Islington Avenue
11      M1B       Scarborough             Rouge

Delete the values where Borough is not assigned

In [3]:
df1 = df['Borough'] != 'Not assigned'
df1.head()

0    False
1    False
2     True
3     True
4     True
Name: Borough, dtype: bool

In [4]:
df1 = df[df['Borough'] != 'Not assigned']
df1.head(10)

Postcode           Borough     Neighbourhood
2       M3A        North York         Parkwoods
3       M4A        North York  Victoria Village
4       M5A  Downtown Toronto      Harbourfront
5       M5A  Downtown Toronto       Regent Park
6       M6A        North York  Lawrence Heights
7       M6A        North York    Lawrence Manor
8       M7A      Queen's Park      Not assigned
10      M9A         Etobicoke  Islington Avenue
11      M1B       Scarborough             Rouge
12      M1B       Scarborough           Malvern

Find all where Neighbourhood is assigned or not and recreate the ones that are not assigned

In [5]:
dfpop = df1[df1['Neighbourhood'] != 'Not assigned']
dfnot = df1[df1['Neighbourhood'] == 'Not assigned']


In [6]:
dfpop.head()

Postcode           Borough     Neighbourhood
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront
5      M5A  Downtown Toronto       Regent Park
6      M6A        North York  Lawrence Heights

In [7]:
dfnot

Postcode       Borough Neighbourhood
8      M7A  Queen's Park  Not assigned

In [8]:
dfnot['Neighbourhood'] =='Queen\'s Park'
    

8    False
Name: Neighbourhood, dtype: bool

In [9]:
dfnow = dfnot.replace(to_replace ="Not assigned", 
                 value ="Queen\'s Park") 

In [10]:
full = dfnow.append(dfpop)

In [11]:
full.head()

Postcode           Borough     Neighbourhood
8      M7A      Queen's Park      Queen's Park
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront
5      M5A  Downtown Toronto       Regent Park

Do a groupby in order to group the neighbourhoods together with a comma separation

In [12]:
df3 =full.groupby(['Postcode','Borough'])['Neighbourhood'].agg([('Neighbourhood', ', '.join)]).reset_index()

In [13]:
df3.head(20)

Postcode      Borough                                      Neighbourhood
0       M1B  Scarborough                                     Rouge, Malvern
1       M1C  Scarborough             Highland Creek, Rouge Hill, Port Union
2       M1E  Scarborough                  Guildwood, Morningside, West Hill
3       M1G  Scarborough                                             Woburn
4       M1H  Scarborough                                          Cedarbrae
5       M1J  Scarborough                                Scarborough Village
6       M1K  Scarborough        East Birchmount Park, Ionview, Kennedy Park
7       M1L  Scarborough                    Clairlea, Golden Mile, Oakridge
8       M1M  Scarborough    Cliffcrest, Cliffside, Scarborough Village West
9       M1N  Scarborough                        Birch Cliff, Cliffside West
10      M1P  Scarborough  Dorset Park, Scarborough Town Centre, Wexford ...
11      M1R  Scarborough                                  Maryvale, Wexford
12      M1S  Scarborough                                          Agincourt
13      M1T  Scarborough            Clarks Corners, Sullivan, Tam O'Shanter
14      M1V  Scarborough  Agincourt North, L'Amoreaux East, Milliken, St...
15      M1W  Scarborough                                    L'Amoreaux West
16      M1X  Scarborough                                        Upper Rouge
17      M2H   North York                                  Hillcrest Village
18      M2J   North York                       Fairview, Henry Farm, Oriole
19      M2K   North York                                    Bayview Village

In [14]:
df3.head()

Postcode      Borough                           Neighbourhood
0      M1B  Scarborough                          Rouge, Malvern
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2      M1E  Scarborough       Guildwood, Morningside, West Hill
3      M1G  Scarborough                                  Woburn
4      M1H  Scarborough                               Cedarbrae

Run a shape function

In [15]:
df3.shape

(103, 3)

# Coursera Applied Data Science: Course 4 Week3 Part2

In [16]:
import pandas as pd
geo = pd.read_csv('Geospatial_Coordinates.csv')
geo.head(12)

Postal Code   Latitude  Longitude
0          M1B  43.806686 -79.194353
1          M1C  43.784535 -79.160497
2          M1E  43.763573 -79.188711
3          M1G  43.770992 -79.216917
4          M1H  43.773136 -79.239476
5          M1J  43.744734 -79.239476
6          M1K  43.727929 -79.262029
7          M1L  43.711112 -79.284577
8          M1M  43.716316 -79.239476
9          M1N  43.692657 -79.264848
10         M1P  43.757410 -79.273304
11         M1R  43.750072 -79.295849

In [17]:
base1=pd.merge(df3, geo, left_on='Postcode', right_on='Postal Code', how='left')
base1.head()

Postcode      Borough                           Neighbourhood Postal Code  \
0      M1B  Scarborough                          Rouge, Malvern         M1B   
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union         M1C   
2      M1E  Scarborough       Guildwood, Morningside, West Hill         M1E   
3      M1G  Scarborough                                  Woburn         M1G   
4      M1H  Scarborough                               Cedarbrae         M1H   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476

In [18]:
base1[base1['Borough'].str.contains('Toronto')]

Postcode           Borough  \
37      M4E      East Toronto   
41      M4K      East Toronto   
42      M4L      East Toronto   
43      M4M      East Toronto   
44      M4N   Central Toronto   
45      M4P   Central Toronto   
46      M4R   Central Toronto   
47      M4S   Central Toronto   
48      M4T   Central Toronto   
49      M4V   Central Toronto   
50      M4W  Downtown Toronto   
51      M4X  Downtown Toronto   
52      M4Y  Downtown Toronto   
53      M5A  Downtown Toronto   
54      M5B  Downtown Toronto   
55      M5C  Downtown Toronto   
56      M5E  Downtown Toronto   
57      M5G  Downtown Toronto   
58      M5H  Downtown Toronto   
59      M5J  Downtown Toronto   
60      M5K  Downtown Toronto   
61      M5L  Downtown Toronto   
63      M5N   Central Toronto   
64      M5P   Central Toronto   
65      M5R   Central Toronto   
66      M5S  Downtown Toronto   
67      M5T  Downtown Toronto   
68      M5V  Downtown Toronto   
69      M5W  Downtown Toronto   
70      M5X  Downtown Toronto   
75      M6G  Downtown Toronto   
76      M6H      West Toronto   
77      M6J      West Toronto   
78      M6K      West Toronto   
82      M6P      West Toronto   
83      M6R      West Toronto   
84      M6S      West Toronto   
87      M7Y      East Toronto   

                                        Neighbourhood Postal Code   Latitude  \
37                                        The Beaches         M4E  43.676357   
41                       The Danforth West, Riverdale         M4K  43.679557   
42                     The Beaches West, India Bazaar         M4L  43.668999   
43                                    Studio District         M4M  43.659526   
44                                      Lawrence Park         M4N  43.728020   
45                                   Davisville North         M4P  43.712751   
46                                 North Toronto West         M4R  43.715383   
47                                         Davisville         M4S  43.704324   
48                        Moore Park, Summerhill East         M4T  43.689574   
49  Deer Park, Forest Hill SE, Rathnelly, South Hi...         M4V  43.686412   
50                                           Rosedale         M4W  43.679563   
51                        Cabbagetown, St. James Town         M4X  43.667967   
52                               Church and Wellesley         M4Y  43.665860   
53                          Harbourfront, Regent Park         M5A  43.654260   
54                           Ryerson, Garden District         M5B  43.657162   
55                                     St. James Town         M5C  43.651494   
56                                        Berczy Park         M5E  43.644771   
57                                 Central Bay Street         M5G  43.657952   
58                           Adelaide, King, Richmond         M5H  43.650571   
59  Harbourfront East, Toronto Islands, Union Station         M5J  43.640816   
60           Design Exchange, Toronto Dominion Centre         M5K  43.647177   
61                     Commerce Court, Victoria Hotel         M5L  43.648198   
63                                           Roselawn         M5N  43.711695   
64                Forest Hill North, Forest Hill West         M5P  43.696948   
65                The Annex, North Midtown, Yorkville         M5R  43.672710   
66                     Harbord, University of Toronto         M5S  43.662696   
67          Chinatown, Grange Park, Kensington Market         M5T  43.653206   
68  CN Tower, Bathurst Quay, Island airport, Harbo...         M5V  43.628947   
69                    Stn A PO Boxes 25 The Esplanade         M5W  43.646435   
70             First Canadian Place, Underground city         M5X  43.648429   
75                                           Christie         M6G  43.669542   
76                       Dovercourt Village, Dufferin         M6H  43.669005   
77                           Little Portugal, Trinity         M6J  43.64792

# Coursera Applied Data Science: Course 4 Week3 Part3

Note to marker: dividing this Part into a few sections to make it more readible

# SECTION1 - drawing data on a map

In [19]:
import pandas as pd
import matplotlib.pyplot as plt
import geopandas as gpd
import folium

In [20]:
#Geometry = [Point(xy) for xy in zip(base1['Longitude'], base1['Latitude'])]

In [21]:
lat1 = list(base1["Latitude"])
lon1 = list(base1["Longitude"])
# cluster = list(base1["cluster"])


In [22]:
tor_map_subset = folium.Map(location=[43.6532, -79.3832]
                    , zoom_start = 12
                    ,tiles='cartodbpositron'    #this is a more minimalist-looking map than the default
                   )

In [23]:
point = folium.map.FeatureGroup()
for lt, ln in zip(lat1, lon1):
    point.add_child(folium.CircleMarker([lt,ln], radius=1, color="red",fill_color="Red"))

tor_map_subset.add_child(point)
tor_map_subset

In [24]:
lat1 = list(base1["Latitude"])
lon1 = list(base1["Longitude"])
Borough = list(base1["Borough"])

tor_map_subset = folium.Map(location=[43.6532, -79.3832]
                    , zoom_start = 12
                    ,tiles='cartodbpositron'    #this is a more minimalist-looking map than the default
                   )

In [25]:

#definition of the condition for color coding
def color_producer(Borough):
    if Borough == 'East Toronto':
        return 'green'
    elif Borough == 'Central Toronto': 
        return 'orange'
    elif Borough == 'Downtown Toronto': 
        return 'blue'
    elif Borough == 'West Toronto': 
        return 'yellow'
    else:
        return 'red' 

In [26]:
points = folium.map.FeatureGroup()
for lt, ln, Borough in zip(lat1, lon1, Borough):
    points.add_child(folium.CircleMarker([lt,ln], radius=1, color=color_producer(Borough),fill_color=color_producer(Borough)))

tor_map_subset.add_child(points)



Observations:
   
    1) Green: East Toronto - the 5 Bouroughs have 7 neighbourboods and located mostly along the coastline
    2) Orange: Central - 9 boroughs mostly with only 1 neighbourhood
    3) Blue: Downtown Toronto is very densely populated
    4) Yellow: West - most of the western Toronto Boroughs have 2 neighbourhoods. Similar to East Toronto where total boroughs here are 8
    5) Other: Red - the other borroughs lie more in the outskirts of Toronto
    

# SECTION 2 - Finding Foursquare data for the points on the map

In [63]:
#from foursuqare
fs = pd.read_csv('toronoto_venues.csv')
fs.head(12)

Unnamed: 0                  Neighborhood  Neighborhood Latitude  \
0            0                   The Beaches              43.676357   
1            1                   The Beaches              43.676357   
2            2                   The Beaches              43.676357   
3            3                   The Beaches              43.676357   
4            4  The Danforth West, Riverdale              43.679557   
5            5  The Danforth West, Riverdale              43.679557   
6            6  The Danforth West, Riverdale              43.679557   
7            7  The Danforth West, Riverdale              43.679557   
8            8  The Danforth West, Riverdale              43.679557   
9            9  The Danforth West, Riverdale              43.679557   
10          10  The Danforth West, Riverdale              43.679557   
11          11  The Danforth West, Riverdale              43.679557   

    Neighborhood Longitude                               Venue  \
0               -79.293031                   Glen Manor Ravine   
1               -79.293031  The Big Carrot Natural Food Market   
2               -79.293031                 Grover Pub and Grub   
3               -79.293031                       Upper Beaches   
4               -79.352188                            Pantheon   
5               -79.352188                        Dolce Gelato   
6               -79.352188                       MenEssentials   
7               -79.352188                     Cafe Fiorentina   
8               -79.352188                          La Diperie   
9               -79.352188                               Mezes   
10              -79.352188             Messini Authentic Gyros   
11              -79.352188              Louis Cifer Brew Works   

    Venue Latitude  Venue Longitude      Venue Category  
0        43.676821       -79.293942               Trail  
1        43.678879       -79.297734   Health Food Store  
2        43.679181       -79.297215                 Pub  
3        43.680563       -79.292869        Neighborhood  
4        43.677621       -79.351434    Greek Restaurant  
5        43.677773       -79.351187      Ice Cream Shop  
6        43.677820       -79.351265      Cosmetics Shop  
7        43.677743       -79.350115  Italian Restaurant  
8        43.677530       -79.352295      Ice Cream Shop  
9        43.677962       -79.350196    Greek Restaurant  
10       43.677827       -79.350569    Greek Restaurant  
11       43.677663       -79.351313             Brewery

In [64]:
from sklearn.cluster import KMeans

In [65]:
# set number of clusters
kclusters = 20

fs_group = fs.drop('Neighborhood', 1)

# run k-means clustering
#kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(fs_group)


In [66]:
fs.groupby('Neighborhood').count()

Unnamed: 0  \
Neighborhood                                                     
Adelaide, King, Richmond                                   100   
Berczy Park                                                 55   
Brockton, Exhibition Place, Parkdale Village                21   
Business Reply Mail Processing Centre 969 Eastern           19   
CN Tower, Bathurst Quay, Island airport, Harbou...          15   
Cabbagetown, St. James Town                                 46   
Central Bay Street                                          88   
Chinatown, Grange Park, Kensington Market                  100   
Christie                                                    15   
Church and Wellesley                                        87   
Commerce Court, Victoria Hotel                             100   
Davisville                                                  38   
Davisville North                                             9   
Deer Park, Forest Hill SE, Rathnelly, South Hil...          14   
Design Exchange, Toronto Dominion Centre                   100   
Dovercourt Village, Dufferin                                20   
First Canadian Place, Underground city                     100   
Forest Hill North, Forest Hill West                          4   
Harbord, University of Toronto                              34   
Harbourfront East, Toronto Islands, Union Station          100   
Harbourfront, Regent Park                                   48   
High Park, The Junction South                               23   
Lawrence Park                                                3   
Little Portugal, Trinity                                    62   
Moore Park, Summerhill East                                  2   
North Toronto West                                          16   
Parkdale, Roncesvalles                                      15   
Rosedale                                                     4   
Roselawn                                                     2   
Runnymede, Swansea                                          38   
Ryerson, Garden District                                   100   
St. James Town                                             100   
Stn A PO Boxes 25 The Esplanade                             94   
Studio District                                             38   
The Annex, North Midtown, Yorkville                         23   
The Beaches                                                  4   
The Beaches West, India Bazaar                              18   
The Danforth West, Riverdale                                42   

                                                    Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                              100   
Berczy Park                                                            55   
Brockton, Exhibition Place, Parkdale Village                           21   
Business Reply Mail Processing Centre 969 Eastern                      19   
CN Tower, Bathurst Quay, Island airport, Harbou...                     15   
Cabbagetown, St. James Town                                            46   
Central Bay Street                                                     88   
Chinatown, Grange Park, Kensington Market                             100   
Christie                                                               15   
Church and Wellesley                                                   87   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             38   
Davisville North                                                        9   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                     14   
Design Exchange, Toronto Dominion Centre                              100   
Dovercourt Village, Dufferin                                           20   
First Canadian Place, Und

In [67]:
fs.groupby('Venue Category').count()

Unnamed: 0  Neighborhood  \
Venue Category                                            
Adult Boutique                          1             1   
Afghan Restaurant                       1             1   
Airport                                 1             1   
Airport Food Court                      1             1   
Airport Gate                            1             1   
Airport Lounge                          2             2   
Airport Service                         2             2   
Airport Terminal                        2             2   
American Restaurant                    23            23   
Antique Shop                            3             3   
Aquarium                                5             5   
Art Gallery                            12            12   
Art Museum                              2             2   
Arts & Crafts Store                     4             4   
Asian Restaurant                        8             8   
Auto Workshop                           1             1   
BBQ Joint                               2             2   
Baby Store                              1             1   
Bagel Shop                              6             6   
Bakery                                 45            45   
Bank                                    8             8   
Bar                                    39            39   
Baseball Stadium                        2             2   
Basketball Stadium                      4             4   
Beach                                   1             1   
Bed & Breakfast                         1             1   
Beer Bar                               17            17   
Beer Store                              3             3   
Belgian Restaurant                      3             3   
Bistro                                  4             4   
...                                   ...           ...   
Sports Bar                              6             6   
Stadium                                 1             1   
Stationery Store                        1             1   
Steakhouse                             20            20   
Strip Club                              1             1   
Supermarket                             4             4   
Sushi Restaurant                       20            20   
Swim School                             1             1   
Taco Place                              2             2   
Tailor Shop                             4             4   
Taiwanese Restaurant                    1             1   
Tanning Salon                           1             1   
Tapas Restaurant                        1             1   
Tea Room                               12            12   
Tennis Court                            1             1   
Thai Restaurant                        16            16   
Theater                                12            12   
Theme Restaurant                        1             1   
Thrift / Vintage Store                  1             1   
Toy / Game Store                        3             3   
Trail                                   4             4   
Train Station                           3             3   
Vegetarian / Vegan Restaurant          20            20   
Video Game Store                        3             3   
Video Store                             1             1   
Vietnamese Restaurant                   8             8   
Wine Bar                                9             9   
Wings Joint                             1             1   
Women's Store                           1             1   
Yoga Studio                             9             9   

                               Neighborhood Latitude  Neighborhood Longitude  \
Venue Category                                                                 
Adult Boutique                                     1                       1   
Afghan Restaurant                                  1                     

In [68]:
print('There are {} unique categories.'.format(len(fs['Venue Category'].unique())))

There are 237 unique categories.


In [69]:
# one hot encoding
fs_onehot = pd.get_dummies(fs[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
fs_onehot['Neighborhood'] = fs['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [fs_onehot.columns[-1]] + list(fs_onehot.columns[:-1])
fs_onehot = fs_onehot[fixed_columns]

fs_onehot.head()

Yoga Studio  Adult Boutique  Afghan Restaurant  Airport  \
0            0               0                  0        0   
1            0               0                  0        0   
2            0               0                  0        0   
3            0               0                  0        0   
4            0               0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Toy / Game Store  Trail  \
0                 0                    0  ...                 0      1   
1                 0                    0  ...                 0      0   
2                 0                    0  ...                 0      0   
3                 0                    0  ...                 0      0   
4                 0                    0  ...                 0      0   

   Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0              0                              0                 0   
1              0                              0                 0   
2              0                              0                 0   
3              0                              0                 0   
4              0                              0                 0   

   Video Store  Vietnamese Restaurant  Wine Bar  Wings Joint  Women's Store  
0            0                      0         0            0              0  
1            0                      0         0            0              0  
2            0                      0         0            0              0  
3            0                      0         0            0              0  
4            0                      0         0            0              0  

[5 rows x 237 columns]

In [70]:
fs_grouped = fs_onehot.groupby('Neighborhood').mean().reset_index()
fs_grouped.head()

Neighborhood  Yoga Studio  \
0                           Adelaide, King, Richmond     0.000000   
1                                        Berczy Park     0.000000   
2       Brockton, Exhibition Place, Parkdale Village     0.047619   
3  Business Reply Mail Processing Centre 969 Eastern     0.052632   
4  CN Tower, Bathurst Quay, Island airport, Harbo...     0.000000   

   Adult Boutique  Afghan Restaurant   Airport  Airport Food Court  \
0             0.0                0.0  0.000000            0.000000   
1             0.0                0.0  0.000000            0.000000   
2             0.0                0.0  0.000000            0.000000   
3             0.0                0.0  0.000000            0.000000   
4             0.0                0.0  0.066667            0.066667   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  ...  \
0      0.000000        0.000000         0.000000          0.000000  ...   
1      0.000000        0.000000         0.000000          0.000000  ...   
2      0.000000        0.000000         0.000000          0.000000  ...   
3      0.000000        0.000000         0.000000          0.000000  ...   
4      0.066667        0.133333         0.133333          0.133333  ...   

   Toy / Game Store  Trail  Train Station  Vegetarian / Vegan Restaurant  \
0               0.0    0.0            0.0                       0.010000   
1               0.0    0.0            0.0                       0.018182   
2               0.0    0.0            0.0                       0.000000   
3               0.0    0.0            0.0                       0.000000   
4               0.0    0.0            0.0                       0.000000   

   Video Game Store  Video Store  Vietnamese Restaurant  Wine Bar  \
0               0.0          0.0                    0.0      0.01   
1               0.0          0.0                    0.0      0.00   
2               0.0          0.0                    0.0      0.00   
3               0.0          0.0                    0.0      0.00   
4               0.0          0.0                    0.0      0.00   

   Wings Joint  Women's Store  
0          0.0            0.0  
1          0.0            0.0  
2          0.0            0.0  
3          0.0            0.0  
4          0.0            0.0  

[5 rows x 237 columns]

In [71]:
#print each neighbourhood with their top 5 venues
num_top_venues = 5

for hood in fs_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = fs_grouped[fs_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
                 venue  freq
0          Coffee Shop  0.05
1                 Café  0.05
2           Steakhouse  0.04
3  American Restaurant  0.04
4                  Bar  0.04


----Berczy Park----
                venue  freq
0         Coffee Shop  0.09
1        Cocktail Bar  0.05
2              Bakery  0.04
3  Italian Restaurant  0.04
4      Farmers Market  0.04


----Brockton, Exhibition Place, Parkdale Village----
            venue  freq
0  Breakfast Spot  0.10
1            Café  0.10
2     Coffee Shop  0.10
3     Yoga Studio  0.05
4             Gym  0.05


----Business Reply Mail Processing Centre 969 Eastern----
              venue  freq
0       Yoga Studio  0.05
1     Auto Workshop  0.05
2              Park  0.05
3        Comic Shop  0.05
4  Recording Studio  0.05


----CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara----
              venue  freq
0  Airport Terminal  0.13
1    Airport Lounge

In [72]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [73]:
import numpy as np
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = fs_grouped['Neighborhood']

for ind in np.arange(fs_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(fs_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2       Brockton, Exhibition Place, Parkdale Village           Coffee Shop   
3  Business Reply Mail Processing Centre 969 Eastern           Yoga Studio   
4  CN Tower, Bathurst Quay, Island airport, Harbo...        Airport Lounge   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café            Steakhouse                   Bar   
1          Cocktail Bar    Seafood Restaurant                  Café   
2                  Café        Breakfast Spot           Yoga Studio   
3         Auto Workshop           Pizza Place  Gym / Fitness Center   
4      Airport Terminal       Airport Service         Boat or Ferry   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0   American Restaurant       Thai Restaurant          Burger Joint   
1           Cheese Shop        Farmers Market              Beer Bar   
2    Italian Restaurant             Pet Store          Climbing Gym   
3      Recording Studio            Restaurant               Butcher   
4      Sculpture Garden                   Bar              Boutique   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Restaurant        Cosmetics Shop                    Gym  
1            Steakhouse                Bakery     Italian Restaurant  
2            Restaurant  Caribbean Restaurant          Burrito Place  
3         Burrito Place               Brewery             Skate Park  
4                 Plane          Airport Gate     Airport Food Court

# SECTION 3 - Doing k-means clustering

In [74]:
# set number of clusters
kclusters = 10

fs_grouped_clustering = fs_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(fs_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 1, 0, 9, 1, 1, 1, 1, 1])

In [75]:

#rename field
#base1['Neighborhood'] = base1['Neighbourhood']
#base1.head()

In [76]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

fs_merged = fs

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
fs_merged = fs_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

fs_merged.head() # check the last columns!

Unnamed: 0                  Neighborhood  Neighborhood Latitude  \
0           0                   The Beaches              43.676357   
1           1                   The Beaches              43.676357   
2           2                   The Beaches              43.676357   
3           3                   The Beaches              43.676357   
4           4  The Danforth West, Riverdale              43.679557   

   Neighborhood Longitude                               Venue  Venue Latitude  \
0              -79.293031                   Glen Manor Ravine       43.676821   
1              -79.293031  The Big Carrot Natural Food Market       43.678879   
2              -79.293031                 Grover Pub and Grub       43.679181   
3              -79.293031                       Upper Beaches       43.680563   
4              -79.352188                            Pantheon       43.677621   

   Venue Longitude     Venue Category  Cluster Labels 1st Most Common Venue  \
0       -79.293942              Trail               8                 Trail   
1       -79.297734  Health Food Store               8                 Trail   
2       -79.297215                Pub               8                 Trail   
3       -79.292869       Neighborhood               8                 Trail   
4       -79.351434   Greek Restaurant               1      Greek Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   Pub     Health Food Store               Dog Run   
1                   Pub     Health Food Store               Dog Run   
2                   Pub     Health Food Store               Dog Run   
3                   Pub     Health Food Store               Dog Run   
4           Coffee Shop        Ice Cream Shop    Italian Restaurant   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Doner Restaurant            Donut Shop   Dumpling Restaurant   
1        Doner Restaurant            Donut Shop   Dumpling Restaurant   
2        Doner Restaurant            Donut Shop   Dumpling Restaurant   
3        Doner Restaurant            Donut Shop   Dumpling Restaurant   
4  Furniture / Home Store           Yoga Studio           Pizza Place   

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0  Eastern European Restaurant     Electronics Store          Women's Store  
1  Eastern European Restaurant     Electronics Store          Women's Store  
2  Eastern European Restaurant     Electronics Store          Women's Store  
3  Eastern European Restaurant     Electronics Store          Women's Store  
4                      Brewery       Bubble Tea Shop                   Café

In [78]:
fs_merged

Unnamed: 0                                       Neighborhood  \
0              0                                        The Beaches   
1              1                                        The Beaches   
2              2                                        The Beaches   
3              3                                        The Beaches   
4              4                       The Danforth West, Riverdale   
5              5                       The Danforth West, Riverdale   
6              6                       The Danforth West, Riverdale   
7              7                       The Danforth West, Riverdale   
8              8                       The Danforth West, Riverdale   
9              9                       The Danforth West, Riverdale   
10            10                       The Danforth West, Riverdale   
11            11                       The Danforth West, Riverdale   
12            12                       The Danforth West, Riverdale   
13            13                       The Danforth West, Riverdale   
14            14                       The Danforth West, Riverdale   
15            15                       The Danforth West, Riverdale   
16            16                       The Danforth West, Riverdale   
17            17                       The Danforth West, Riverdale   
18            18                       The Danforth West, Riverdale   
19            19                       The Danforth West, Riverdale   
20            20                       The Danforth West, Riverdale   
21            21                       The Danforth West, Riverdale   
22            22                       The Danforth West, Riverdale   
23            23                       The Danforth West, Riverdale   
24            24                       The Danforth West, Riverdale   
25            25                       The Danforth West, Riverdale   
26            26                       The Danforth West, Riverdale   
27            27                       The Danforth West, Riverdale   
28            28                       The Danforth West, Riverdale   
29            29                       The Danforth West, Riverdale   
...          ...                                                ...   
1667        1667                                 Runnymede, Swansea   
1668        1668                                 Runnymede, Swansea   
1669        1669                                 Runnymede, Swansea   
1670        1670                                 Runnymede, Swansea   
1671        1671                                 Runnymede, Swansea   
1672        1672                                 Runnymede, Swansea   
1673        1673                                 Runnymede, Swansea   
1674        1674                                 Runnymede, Swansea   
1675        1675                                 Runnymede, Swansea   
1676        1676                                 Runnymede, Swansea   
1677        1677                                 Runnymede, Swansea   
1678        1678  Business Reply Mail Processing Centre 969 Eastern   
1679        1679  Business Reply Mail Processing Centre 969 Eastern   
1680        1680  Business Reply Mail Processing Centre 969 Eastern   
1681        1681  Business Reply Mail Processing Centre 969 Eastern   
1682        1682  Business Reply Mail Processing Centre 969 Eastern   
1683        1683  Business Reply Mail Processing Centre 969 Eastern   
1684        1684  Business Reply Mail Processing Centre 969 Eastern   
1685        1685  Business Reply Mail Processing Centre 969 Eastern   
1686        1686  Business Reply Mail Processing Centre 969 Eastern   
1687        1687  Business Reply Mail Processing Centre 969 Eastern   
1688        1688  Business Reply Mail Processing Centre 969 Eastern   
1689        1689  Business Reply Mail Processing Centre 969 Eastern   
1690        1690  Business Reply Mail Processing Centre 969 Eastern   
1691        1691  Business Rep

In [79]:
import matplotlib.cm as cm
import matplotlib.colors as colors
# create map
map_clusters = folium.Map(location=[43.6532, -79.3832], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(fs_merged['Neighborhood Latitude'], fs_merged['Neighborhood Longitude'], fs_merged['Neighborhood'], fs_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [91]:
clustergr1 = fs_merged[['Cluster Labels','Neighborhood']]
clustergr1.head()

Cluster Labels                  Neighborhood
0               8                   The Beaches
1               8                   The Beaches
2               8                   The Beaches
3               8                   The Beaches
4               1  The Danforth West, Riverdale

In [92]:
clustergr = clustergr1.groupby('Cluster Labels').count()
clustergr

Neighborhood
Cluster Labels              
0                         19
1                       1635
2                          4
3                          2
4                          3
5                          2
6                          9
7                          4
8                          4
9                         15

# SECTION 4 - Observations

Observations SECTION1:
   
    1) Green: East Toronto - the 5 Bouroughs have 7 neighbourboods and located mostly along the coastline
    2) Orange: Central - 9 boroughs mostly with only 1 neighbourhood
    3) Blue: Downtown Toronto is very densely populated
    4) Yellow: West - most of the western Toronto Boroughs have 2 neighbourhoods. Similar to East Toronto where total boroughs here are 8
    5) Other: Red - the other borroughs lie more in the outskirts of Toronto
    
Observations SECTION2:  
    
Using Foursquare data for venues around all Neighbourhoods.
There are 237 unique categories of venues accross the Neighbourhoods.
    
Observations SECTION3:
    
    1) Divided the data into 10 clusters, but most data seem to fall together into the same cluster (cluster1 as seen above)
    2) This means that the data could be divided into less clusters (perhaps 3)
    3) It shows that the areas are quite similar in nature